### PR TITLE
Validate path, query and request parameters separately

### PR DESCRIPTION
**Pull request by quentindemetz**
fixes https://github.com/interagent/committee/issues/339
fixes https://github.com/interagent/committee/issues/328

### CODE REVIEW

**Comment by quentindemetz** [comment]
👋🏻

File: lib/committee/schema_validator/open_api_3/operation_wrapper.rb [hunk 1]
```diff
@@ -74,53 +74,58 @@ def request_content_types
...
-          request_operation.validate_request_body(content_type, params, schema_validator_options)
+          validate_path_and_query_params(path_params, query_params, headers, validator_option)
+          request_operation.validate_request_body(content_type, body_params, build_openapi_parser_body_option(validator_option))
         rescue => e
           raise Committee::InvalidRequest.new(e.message, original_error: e)
         end
 
+        def validate_path_and_query_params(path_params, query_params, headers, validator_option)
+          # it's currently impossible to validate path params and query params separately
+          # so we have to resort to this workaround
```
Comment (ota42y) [inline lib/committee/schema_validator/open_api_3/operation_wrapper.rb:114]: openapi_parser have validate_path_params method.
https://github.com/ota42y/openapi_parser/blob/f6a975791fc58b19d40498daeb88b80e01c70c16/lib/openapi_parser/request_operation.rb#L42
That should have made it compatible with separate validation... I'll do check that code 🙋

File: lib/committee/schema_validator/open_api_3.rb [hunk 2]
```diff
@@ -80,24 +92,16 @@ def request_unpack(request)
...
         query_param = unpacker.unpack_query_params(request)
+        query_param.merge!(request_param) if request.get? && validator_option.allow_get_body
+        request.env[validator_option.query_hash_key] = query_param
+      end
 
+      def copy_coerced_data_to_params(request)
         request.env[validator_option.params_key] = Committee::Utils.indifferent_hash
-        request.env[validator_option.params_key].merge!(Committee::Utils.deep_copy(query_param))
         request.env[validator_option.params_key].merge!(Committee::Utils.deep_copy(request.env[validator_option.request_body_hash_key]))
+        request.env[validator_option.params_key].merge!(Committee::Utils.deep_copy(request.env[validator_option.query_hash_key]))
```
Comment (ota42y) [inline lib/committee/schema_validator/open_api_3.rb:102]: I think it's very good to prioritize path, query, and body in that order, just like Rails.
But this is breaking changes so I add backward compatibility code for the users which want to old priority.
(This is memo. You do not need to fix it :) )

**Review by ota42y — COMMENTED** [review]
This is very good code!!!!
But there was one poit that needed to be investigated, so I'll merge after it's done.

**Comment by JohannesLamberts** [comment]
@ota42y we are currently impacted by #328 as well and would be interested in a fix. Is it possible to support solving the issue / MR?

**Comment by krutovgerman** [comment]
Hey people, would you accept this PR? We are waiting for it too. Pleaaaaseeeee!

**Comment by klefix** [comment]
Hey, we are still affected by this issue and would not like to fork this repo just to get this merged. 
Is there anything missing in this PR?

File: lib/committee/schema_validator/open_api_3/operation_wrapper.rb (hunk 1 above)
Comment (ota42y) [reply lib/committee/schema_validator/open_api_3/operation_wrapper.rb:114]: openapi_parser have validate_path_params but operation_wrapper doesn't delegate it so we can't use it.
Ok, this code is no problem 🙆‍♀️

**Review by ota42y — APPROVED** [review]

**Comment by ota42y** [comment]
Sorry for the very delay, thanks for the changes!

**Comment by ota42y** [comment]
I released 5.0.0 with this changes!
https://rubygems.org/gems/committee/versions/5.0.0

**Pull request merged**